Add operation with wedge to ease the representation of the topology of core mesh, with wedges to handle attribs.

dlyr opened the pull request:
* **Please check if the PR fulfills these requirements**

First part of this PR has been moved to #595 (add quad meshes for `Core` and `Engine`).
Second part of this PR has been moved to #719  (add wedge collapse).

Next steps, sorted by decreasing priority:
- [ ] Topological test collapse on non triangular faces
- [ ] Topological mesh for polymesh: split
- [ ] Topological mesh add/del face/halfedge/edge/vertex which should update wedges ref count
- [ ] Add Catmull-Clark subdivision (from previous implementation, see in code todo's)
- [ ] Refactor UpdateGL to remove duplicated code
- [ ] Test on real data
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Check picking/wireframe to show only edges corresponding to the core geometry
- [ ] Cleanup history __NO FORCE PUSH IN THE MEANTIME__ (other than rebase)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Add operation with wedge to ease the representation of the topology of core mesh, with wedges to handle attribs.

nmellado commented on src/Core/Geometry/TriangleMesh.hpp:
```diff
@@ -249,6 +250,13 @@ class IndexedGeometry : public AttribArrayGeometry
     /// \warning There is no error check on the handles attribute type.
     inline bool append( const IndexedGeometry<IndexType>& other );
 
+    const IndexContainerType& getIndices() const { return m_indices; }
```
Other changes suggested:
 - mark `m_indices` as private
 - add access method `withLock` and `unlock`

dlyr replied:
Yes, we should do it for this PR (but it will change a lot of file, let's postpone when the PR converges).

nmellado reviewed (commented):
Nice !

MathiasPaulin commented on src/Core/Geometry/TriangleMesh.hpp:
```diff
@@ -249,6 +250,13 @@ class IndexedGeometry : public AttribArrayGeometry
+        m_indices = indices;
+        notify();
+    }
     ///\todo make it protected
```
protected data might prevent notifiction and invariant maintenance.
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rh-protected
prefer make it private

MathiasPaulin commented on src/Engine/Renderer/Mesh/Mesh.inl:
```diff
@@ -464,10 +467,62 @@ PointCloud::PointCloud( const std::string& name,
+        {
+            auto i0 {face[0]};
+            auto i1 {face[1]};
+            /// simple fan type triangulation
```
Should be better to use sewing triangulation, that is the same as fan one for polygon with less than 
5 vertices but will prevent rendering artifacts for polygons with more than 6 vertices while being as simpler as fan triangulation for convex planar polygons.
![IMG_1340](https://user-images.githubusercontent.com/11328578/87122934-1a3f8c80-c286-11ea-9d82-97ffbe430797.jpg)

nmellado replied:
Do you have a specific algorithm in mind ?

MathiasPaulin replied:
Yes, I draw it in red on the picture above. It is really very similar to the fan algo 
Instead of generating list of visited vertices (and emit triangle at each visited vertice starting at the third) in a fan mode (e.g 0123456 in the above example), you generate it in a strip mode (0126354) and emit a triangle at for each visited vertex starting at the third. Then, the only difference between the two is just the order of the visited vertices.
Not so clear if explained like this but I can implement it rather quickly (well, it might take 1 hour if if find such an hour in my agenda :p )

dlyr replied:
That's fine I'll do it

dlyr replied:
first version done, but I'll need to simplify it.

MathiasPaulin replied:
35 min left for the "one hour implementation". You're in time. Run @dlyr, run :D

dlyr replied:
well I've finish my second one...

dlyr replied:
But still want to get rid of one index ...

hoshiryu replied:
There is some kind of related algorithm used when removing an edge in a tetrahedral mesh, which is done by dynamic programming (see paper `Minimal  Triangulations  of  Polygonal  Domains`)

dlyr replied:
I have pushed the two (not tested as everything in the draft PR) versions

MathiasPaulin reviewed (commented):
Nice first shot. Some improvements suggested.

nmellado commented:
Next steps:

- [ ] Makes `m_indices` private
- [ ] Update assimp loader to disable triangulation
- [ ] Update loading process to produce `PolyMesh` when required
- [ ] Test on real data
- [ ] (not sure about that one, but might be necessary for testing) Add Catmull-Clark subdivision (from previous implementation)

MathiasPaulin commented on src/Engine/Renderer/Mesh/Mesh.inl:
```diff
@@ -464,10 +467,83 @@ PointCloud::PointCloud( const std::string& name,
+
+        if ( face.size() == 3 ) { m_triangleIndices.push_back( face ); }
+        else
+        {
```
Note that both algorithms might fail if some vertices are aligned. Imagine the following rectangular element (where `x` denote a vertex, , the first one being identified as `v0`): 
```
   x---------------------------------------------------------------x
   |                                                               |
   |                                                               |
   x------x---------x----------x---------x-----------x-------------x
v0
```
both version will fail to generate usefull (even non manifold in some cases) triangulations. 
 - Fan will generate degenerates triangles (the first ones are degenerates)
 - Sew will generate overlapping triangle (from the 4th to the last)

Even if this case is very rare, there are others situations of aligned vertices, more frequent, that might fail the same way.

But for a first implementation, that will allow to test the overall behavior, it is fine for me.

Note that, event if triangulating a convex polygon seems simple, there is a huge litterature about how to do that robustly and with minimal complexity as the number of different triangulation of a n-vertex polygon is given by the (n-2) Catalan number (Euler). For the above exemple, with 9 vertices, there is 132 different triangulations (https://en.wikipedia.org/wiki/Polygon_triangulation)

MathiasPaulin replied:
I agree, who cares about such malformed polygons ;)

dlyr replied:
well, I'll switch back to the triangle fan implementation.
Here the purpose is to handle quad to triangle conversion on the fly, not to triangulate nicely the mesh.
If you want to do better, use openmesh's, or assimp's functionality. 
As said in the wikipedia page you link "It is trivial to triangulate any convex polygon in linear time into a fan triangulation, by adding diagonals from one vertex to all other vertices."
If you want to spend time on the PR, could you please focus on the PolyMesh to TopoMesh conversion and tests ?

dlyr replied:
(BTW could you draw how sew creates overlapping triangle, on my side I found degenerated, but no overlapping)

nmellado replied:
What will be the impact of these degenerated triangles in the rendering ?

nmellado replied:
(And also: 99% of our cases will planar quads, so we don’t really care).

MathiasPaulin replied:
sew (strip) is nice for $99.\bar{9}$ of our triangle, we'll keep this.
An yes, I'll focus on PolyMesh to TopoMesh conversion ;)

nmellado commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176&bid=19090247)

nmellado commented on tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp:
```diff
@@ -477,6 +478,58 @@ void MinimalComponent::initialize() {
+
+        } );
+        auto quad = VectorNui( 4 );
+        quad << 0, 1, 2, 3;
```
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

dlyr replied:
Please tell eigen :D

nmellado commented on tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp:
```diff
@@ -477,6 +478,58 @@ void MinimalComponent::initialize() {
+        auto hepta = VectorNui( 7 );
+        hepta << 3, 2, 4, 5, 6, 7, 8;
+        auto degen = VectorNui( 10 );
+        degen << 1, 0, 9, 10, 11, 12, 13, 14, 15, 16;
```
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado commented on tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp:
```diff
@@ -477,6 +478,58 @@ void MinimalComponent::initialize() {
+        auto quad = VectorNui( 4 );
+        quad << 0, 1, 2, 3;
+        auto hepta = VectorNui( 7 );
+        hepta << 3, 2, 4, 5, 6, 7, 8;
```
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado commented on tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp:
```diff
@@ -477,6 +478,58 @@ void MinimalComponent::initialize() {
+        auto degen = VectorNui( 10 );
+        degen << 1, 0, 9, 10, 11, 12, 13, 14, 15, 16;
+        auto degen2 = VectorNui( 10 );
+        degen2 << 14, 13, 12, 11, 10, 9, 17, 18, 16, 15;
```
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176&bid=19090247)

nmellado commented on tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp:
```diff
@@ -477,6 +478,58 @@ void MinimalComponent::initialize() {
+
+        } );
+        auto quad = VectorNui( 4 );
+        quad << 0, 1, 2, 3;
```
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

MathiasPaulin commented:
@nmellado the answer to your question about impact of degenerated triangles on rendering is on the picture post by @dlyr.
Seen from here, I'm sure there is no T vertices on the polymesh and everything is well done. clap clap @dlyr.  But I think there are  degenerated triangles along the horizontal edge just abode the bottom. If, instead of rendering flat color, you have to interpolate attributes to compute smooth shading, you'll have interpolation artifacts along that edges. degenerated triangles will be removed by the rasterizer and you will see discontinuity in the attribute interpolation.

dlyr commented:
This kind of artifact.

![Screenshot from 2020-07-10 22-45-52](https://user-images.githubusercontent.com/11328509/87201617-41867000-c2ff-11ea-8a98-a033e970c5d9.png)

Note that it is also present in "simple quads" and are not linked to degenerated triangles, just to thin triangles as seen here:
![Screenshot from 2020-07-10 22-48-16](https://user-images.githubusercontent.com/11328509/87201742-7c88a380-c2ff-11ea-8363-1fe40b48baf9.png)

MathiasPaulin commented:
Thin triangles are bad for rendering. More than computation precisions that degrade as the triangles become thin, the most proeminent problem relies to interpolation along the triangles in a mesh : piecewise linear approximation of non linear function over the mesh exhibit large errors and continuity issues. That's why a good mesh for rendering is a mesh that maximise the minimum triangle area and that maximise also the minimum internal angle on the triangle mesh (Delaunay triangulation does this).

nmellado commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176&bid=19090247)

nmellado commented on tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp:
```diff
@@ -477,6 +478,69 @@ void MinimalComponent::initialize() {
+        auto hepta = VectorNui( 7 );
+        hepta << 3, 2, 4, 5, 6, 7, 8;
+        auto degen = VectorNui( 10 );
+        degen << 1, 0, 9, 10, 11, 12, 13, 14, 15, 16;
```
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado commented on tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp:
```diff
@@ -477,6 +478,69 @@ void MinimalComponent::initialize() {
+        polyMesh.setNormals( normals );
+
+        auto quad = VectorNui( 4 );
+        quad << 0, 1, 2, 3;
```
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado commented on tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp:
```diff
@@ -477,6 +478,69 @@ void MinimalComponent::initialize() {
+        auto degen = VectorNui( 10 );
+        degen << 1, 0, 9, 10, 11, 12, 13, 14, 15, 16;
+        auto degen2 = VectorNui( 10 );
+        degen2 << 14, 13, 12, 11, 10, 9, 17, 18, 16, 15;
```
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado commented on tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp:
```diff
@@ -477,6 +478,69 @@ void MinimalComponent::initialize() {
+        auto quad = VectorNui( 4 );
+        quad << 0, 1, 2, 3;
+        auto hepta = VectorNui( 7 );
+        hepta << 3, 2, 4, 5, 6, 7, 8;
```
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176&bid=19090247)

nmellado commented on tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp:
```diff
@@ -477,6 +478,71 @@ void MinimalComponent::initialize() {
+        auto quad = VectorNui( 4 );
+        quad << 0, 1, 2, 3;
+        auto hepta = VectorNui( 7 );
+        hepta << 3, 2, 4, 5, 6, 7, 8;
```
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado commented on tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp:
```diff
@@ -477,6 +478,71 @@ void MinimalComponent::initialize() {
+        auto degen = VectorNui( 10 );
+        degen << 1, 0, 9, 10, 11, 12, 13, 14, 15, 16;
+        auto degen2 = VectorNui( 10 );
+        degen2 << 14, 13, 12, 11, 10, 9, 17, 18, 16, 15;
```
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado commented on tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp:
```diff
@@ -477,6 +478,71 @@ void MinimalComponent::initialize() {
+        auto hepta = VectorNui( 7 );
+        hepta << 3, 2, 4, 5, 6, 7, 8;
+        auto degen = VectorNui( 10 );
+        degen << 1, 0, 9, 10, 11, 12, 13, 14, 15, 16;
```
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado commented on tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp:
```diff
@@ -477,6 +478,71 @@ void MinimalComponent::initialize() {
+        polyMesh.setNormals( normals );
+
+        auto quad = VectorNui( 4 );
+        quad << 0, 1, 2, 3;
```
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           

Clones added
============
- tests/unittest/Core/topomesh.cpp  2
- tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp  2
         

Clones removed
==============
+ src/Core/Geometry/TopologicalMesh.cpp  -7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176&bid=19090247)

nmellado commented on tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp:
```diff
@@ -477,6 +478,71 @@ void MinimalComponent::initialize() {
+        auto degen = VectorNui( 10 );
+        degen << 1, 0, 9, 10, 11, 12, 13, 14, 15, 16;
+        auto degen2 = VectorNui( 10 );
+        degen2 << 14, 13, 12, 11, 10, 9, 17, 18, 16, 15;
```
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

dlyr replied:
@nmellado could you mark these kind of cppcheck/codacy warning as ignored ? (see [here](https://support.codacy.com/hc/en-us/articles/360009180134-Issues-View), but I do not test it)

nmellado commented on tests/unittest/Core/topomesh.cpp:
```diff
@@ -336,6 +369,67 @@ void test_split( TopologicalMesh& topo, TopologicalMesh::EdgeHandle eh, float f
+    auto quad = VectorNui( 4 );
+    quad << 0, 1, 2, 3;
+    auto hepta = VectorNui( 7 );
+    hepta << 3, 2, 4, 5, 6, 7, 8;
```
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado commented on tests/unittest/Core/topomesh.cpp:
```diff
@@ -336,6 +369,67 @@ void test_split( TopologicalMesh& topo, TopologicalMesh::EdgeHandle eh, float f
+    auto degen = VectorNui( 10 );
+    degen << 1, 0, 9, 10, 11, 12, 13, 14, 15, 16;
+    auto degen2 = VectorNui( 10 );
+    degen2 << 14, 13, 12, 11, 10, 9, 17, 18, 16, 15;
```
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado commented on tests/unittest/Core/topomesh.cpp:
```diff
@@ -336,6 +369,67 @@ void test_split( TopologicalMesh& topo, TopologicalMesh::EdgeHandle eh, float f
+    polyMesh.setNormals( normals );
+
+    auto quad = VectorNui( 4 );
+    quad << 0, 1, 2, 3;
```
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado commented on tests/unittest/Core/topomesh.cpp:
```diff
@@ -336,6 +369,67 @@ void test_split( TopologicalMesh& topo, TopologicalMesh::EdgeHandle eh, float f
+    auto hepta = VectorNui( 7 );
+    hepta << 3, 2, 4, 5, 6, 7, 8;
+    auto degen = VectorNui( 10 );
+    degen << 1, 0, 9, 10, 11, 12, 13, 14, 15, 16;
```
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           

Clones added
============
- tests/unittest/Core/topomesh.cpp  2
- tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp  2
         

Clones removed
==============
+ src/Core/Geometry/TopologicalMesh.cpp  -7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176&bid=19090247)

dlyr committed:
[tests] Remove deprecated topo tests.

remove deprecated tests on normals

nmellado commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           

Clones added
============
- tests/unittest/Core/topomesh.cpp  2
- tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp  2
         

Clones removed
==============
+ src/Core/Geometry/TopologicalMesh.cpp  -10
+ src/Core/Geometry/TopologicalMesh.hpp  -10
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176&bid=19090247)

nmellado commented on tests/unittest/Core/topomesh.cpp:
```diff
@@ -336,6 +361,67 @@ void test_split( TopologicalMesh& topo, TopologicalMesh::EdgeHandle eh, float f
+    auto quad = VectorNui( 4 );
+    quad << 0, 1, 2, 3;
+    auto hepta = VectorNui( 7 );
+    hepta << 3, 2, 4, 5, 6, 7, 8;
```
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado commented on tests/unittest/Core/topomesh.cpp:
```diff
@@ -336,6 +361,67 @@ void test_split( TopologicalMesh& topo, TopologicalMesh::EdgeHandle eh, float f
+    auto degen = VectorNui( 10 );
+    degen << 1, 0, 9, 10, 11, 12, 13, 14, 15, 16;
+    auto degen2 = VectorNui( 10 );
+    degen2 << 14, 13, 12, 11, 10, 9, 17, 18, 16, 15;
```
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado commented on tests/unittest/Core/topomesh.cpp:
```diff
@@ -336,6 +361,67 @@ void test_split( TopologicalMesh& topo, TopologicalMesh::EdgeHandle eh, float f
+    auto hepta = VectorNui( 7 );
+    hepta << 3, 2, 4, 5, 6, 7, 8;
+    auto degen = VectorNui( 10 );
+    degen << 1, 0, 9, 10, 11, 12, 13, 14, 15, 16;
```
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado commented on tests/unittest/Core/topomesh.cpp:
```diff
@@ -336,6 +361,67 @@ void test_split( TopologicalMesh& topo, TopologicalMesh::EdgeHandle eh, float f
+    polyMesh.setNormals( normals );
+
+    auto quad = VectorNui( 4 );
+    quad << 0, 1, 2, 3;
```
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           

Clones added
============
- tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp  2
- tests/unittest/Core/topomesh.cpp  2
         

Clones removed
==============
+ src/Core/Geometry/TopologicalMesh.hpp  -10
+ src/Core/Geometry/TopologicalMesh.cpp  -10
+ src/Engine/Component/GeometryComponent.cpp  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176&bid=19090247)

nmellado commented:
PolyMesh loading has been tested using this file:  [cube_quads.zip](https://github.com/STORM-IRIT/Radium-Engine/files/4946309/cube_quads.zip)

nmellado commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           

Clones added
============
- tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp  2
- tests/unittest/Core/topomesh.cpp  2
         

Clones removed
==============
+ src/Core/Geometry/TopologicalMesh.hpp  -10
+ src/Engine/Component/GeometryComponent.cpp  -6
+ src/Core/Geometry/TopologicalMesh.cpp  -10
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176&bid=19090247)

nmellado commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           

Clones added
============
- tests/unittest/Core/topomesh.cpp  3
- tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp  3
         

Clones removed
==============
+ src/Core/Geometry/TopologicalMesh.hpp  -10
+ src/Engine/Component/GeometryComponent.cpp  -6
+ src/Core/Geometry/TopologicalMesh.cpp  -10
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176&bid=19090247)

nmellado commented on tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp:
```diff
@@ -18,10 +21,56 @@
+    auto quad = VectorType( 4 );
+    quad << 0, 1, 2, 3;
+    auto hepta = VectorType( 7 );
+    hepta << 3, 2, 4, 5, 6, 7, 8;
```
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado commented on tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp:
```diff
@@ -18,10 +21,56 @@
+    auto hepta = VectorType( 7 );
+    hepta << 3, 2, 4, 5, 6, 7, 8;
+    auto degen = VectorType( 10 );
+    degen << 1, 0, 9, 10, 11, 12, 13, 14, 15, 16;
```
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado commented on tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp:
```diff
@@ -18,10 +21,56 @@
+{
+    using VectorType = Eigen::Matrix<uint, Eigen::Dynamic, 1>;
+    auto quad = VectorType( 4 );
+    quad << 0, 1, 2, 3;
```
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado commented on tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp:
```diff
@@ -18,10 +21,56 @@
+    auto degen = VectorType( 10 );
+    degen << 1, 0, 9, 10, 11, 12, 13, 14, 15, 16;
+    auto degen2 = VectorType( 10 );
+    degen2 << 14, 13, 12, 11, 10, 9, 17, 18, 16, 15;
```
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           

Clones added
============
- tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp  3
- tests/unittest/Core/topomesh.cpp  3
         

Clones removed
==============
+ src/Engine/Component/GeometryComponent.cpp  -6
+ src/Core/Geometry/TopologicalMesh.hpp  -10
+ src/Core/Geometry/TopologicalMesh.cpp  -10
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176&bid=19090247)

nmellado commented on tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp:
```diff
@@ -18,10 +21,53 @@
+    auto degen = VectorType( 10 );
+    degen << 1, 0, 9, 10, 11, 12, 13, 14, 15, 16;
+    auto degen2 = VectorType( 10 );
+    degen2 << 14, 13, 12, 11, 10, 9, 17, 18, 16, 15;
```
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado commented on tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp:
```diff
@@ -18,10 +21,53 @@
+Ra::Core::VectorNuArray getPolyMeshFaces() {
+    using VectorType = Eigen::Matrix<uint, Eigen::Dynamic, 1>;
+    auto quad        = VectorType( 4 );
+    quad << 0, 1, 2, 3;
```
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado commented on tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp:
```diff
@@ -18,10 +21,53 @@
+    auto quad        = VectorType( 4 );
+    quad << 0, 1, 2, 3;
+    auto hepta = VectorType( 7 );
+    hepta << 3, 2, 4, 5, 6, 7, 8;
```
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado commented on tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp:
```diff
@@ -18,10 +21,53 @@
+    auto hepta = VectorType( 7 );
+    hepta << 3, 2, 4, 5, 6, 7, 8;
+    auto degen = VectorType( 10 );
+    degen << 1, 0, 9, 10, 11, 12, 13, 14, 15, 16;
```
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado replied:
Codacy found an issue: [Found suspicious operator ','](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176)

nmellado commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           

Clones added
============
- tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp  3
- tests/unittest/Core/topomesh.cpp  3
         

Clones removed
==============
+ src/Core/Geometry/TopologicalMesh.hpp  -10
+ src/Engine/Component/GeometryComponent.cpp  -6
+ src/Core/Geometry/TopologicalMesh.cpp  -10
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176&bid=19090247)

nmellado commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Clones added
============
- tests/unittest/Core/topomesh.cpp  3
- tests/ExampleApps/DrawPrimitivesApp/minimalradium.cpp  3
         

Clones removed
==============
+ src/Core/Geometry/TopologicalMesh.hpp  -10
+ src/Core/Geometry/TopologicalMesh.cpp  -10
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5831176&bid=19090247)

nmellado commented:
@dlyr what is the current status of this PR ? Could we merge some of the functionalities ?

dlyr commented:
Well the status is as describe in the first comment, there is some work to do, and nothing is really tested.
Something that can really help is to remove prop management in topological mesh and use only wedges.

nmellado commented:
Chems do you think you could do that ?

nmellado commented:
@dlyr I have update the PR description to rephrase the current state and next steps to do.

dlyr commented:
This PR contains hot fixes for wedges, if nobody works on hit, I'll do a hot fix PR with just the fixes, and the rest will wait.
@nmellado @loicbarthe ?

nmellado commented:
Yes please, push the hot fix and leave the remaining work to another PR.

dlyr committed:
[core] Remove unneeded lines.

dlyr committed:
[core] Add update from/to TriangleMesh and update wedge normals.

dlyr committed:
[core] Topo: move template specialization to the top.

dlyr committed:
[core] Topo/Core mesh "fast" updates.

dlyr committed:
[core] Topo/Core mesh updates: switch to vector when possible.

dlyr committed:
[core] Fix Topo updates and optimizing, sort of.

dlyr committed:
[core] Topo store normal index, use direct acces to wedge data.

dlyr committed:
[core] Topo reorganize declarations (hpp/inl).

dlyr committed:
[core] remove deprecated unused using decl.

dlyr committed:
[core] rename addProp to addAttrib form wedge collection.

dlyr committed:
[core] rename setWedgeData to setWedgeAttrib to set  one attrib.

dlyr committed:
[core] more deprecation, is "mostly" handled by wedges,

need new helpers function.

dlyr committed:
[core] topo remove omp parallel for (windows error)

dlyr committed:
[core] topo add newWedge function and some aliases.

dlyr committed:
[core] topo improve newWedgeData.

dlyr committed:
[core] Topo to line Mesh.

dlyr committed:
[core] More checks in topo integrity.

dlyr committed:
[core] Add topo triangulate fixup.

dlyr committed:
[tests] Draw primitive triangulate, and remove degen faces.

dlyr committed:
[core] fix topo wedge collapse

core topo use modified version from open mesh collapse.

move to collapse internal to private

collapse fix loop wedge.

detect feature for collapse.

dlyr committed:
[tests] wip topo merge unit test

dlyr committed:
[tests] draw prim wedge collapse

dlyr committed:
[core] fix mergeEqualWedge for boundaries.

dlyr committed:
[tests] DrawPrimitives more examples of collapse.

dlyr committed:
[core] rename addWedgeData to addWedgeAttrib, when one attrib only.

dlyr committed:
[tests] Update DrawPrimitive for collapse.

dlyr committed:
[core] clean and add deprecated.

dlyr committed:
[tests] update topo collapse unittest.

dlyr committed:
[core] save topological mesh current status as deprecated.

dlyr committed:
[core] clean deprecated code that refer new version.

dlyr committed:
[tests] Topo: make conversion from PolyMesh (wedges only).

dlyr committed:
[core] Topo: make conversion from PolyMesh (wedges only).

dlyr committed:
[core] huge remove of deprecated, need to update tests.

dlyr committed:
[core] Topo: Wedge modifier.

dlyr committed:
[tests] call splitEdge instead of alias splitEdgeWedge.

[tests] topo clean test naming.

[tests] add wedge management tests, wip to triangulate.

[tests] Add topo triangulate.

[tests] newWedgeData(he).

[tests] topo unit tests check integrity of triangulated.

dlyr committed:
[core] Make topo ctor templated by index type (tri/poly).

MathiasPaulin commented:
Is this draft PR still needed ?

dlyr commented:
yes, there is all the TODO's to do
